syntax = "proto3";

package raft;

service RaftNode {
  rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
  rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
  rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);
}

message RequestVoteRequest {
  int64 term = 1;
  string candidate_id = 2;
  int64 last_log_index = 3;
  int64 last_log_term = 4;
}

message RequestVoteResponse {
  int64 term = 1;
  bool vote_granted = 2;
}

message AppendEntriesRequest {
  int64 term = 1;
  string leader_id = 2;
  int64 prev_log_index = 3;
  int64 prev_log_term = 4;
  repeated LogEntry entries = 5;
  int64 leader_commit = 6;
}

message AppendEntriesResponse {
  int64 term = 1;
  bool success = 2;
  int64 match_index = 3;
}

message LogEntry {
  int64 term = 1;
  int64 index = 2;
  string command = 3;
  bytes data = 4;
}

message HealthCheckRequest {
  string node_id = 1;
}

message HealthCheckResponse {
  string status = 1;
  string node_id = 2;
  string raft_state = 3;
  int64 current_term = 4;
  int64 commit_index = 5;
  int64 last_applied = 6;
}
